bug #930: added email property check, ensuring the correct response.

dev-rahulbhadoriya opened the pull request:
Closes #930

Added a check for the email field. If the property is missing or not included in the request body, an error 400.23 ( propertyNotFound) will be returned. 

```javascript
{
    "message": "Could not find a property named 'email'",
    "code": 400.23,
    "details": {
        "property": "email"
    }
}
``` 
#### What has been done to verify that this works as intended?

 Tested with postman.

#### Why is this the best possible solution? Were any other approaches considered?

#### How does this change affect users? Describe intentional changes to behavior and behavior that could have accidentally been affected by code changes. In other words, what are the regression risks?

#### Does this change require updates to the API documentation? If so, please update docs/api.md as part of this PR.

#### Before submitting this PR, please make sure you have:

- [x] run `make test-full` and confirmed all checks still pass OR confirm CircleCI build passes
- [ ] verified that any code from external sources are properly credited in comments or that everything is internally sourced

matthew-white commented:
Thanks, @dev-rahulbhadoriya! I have a couple of small suggestions. First, I think we should use `Problem.user.missingParameter` instead of `propertyNotFound`. Currently, we only use `propertyNotFound` when validating the `$select` OData query parameter, where the Problem means "you specified a property that doesn't exist." In other cases, it's more common that we use `missingParameter`, which has more of a meaning of "you didn't specify a property that we expected."

It'd be great if you could also add a test to test/integration/api/users.js for the new behavior. The test could send a request without `email`, then assert that the resulting response body has correct values for `code` and `details`.

Also, just FYI, the project is going to be quieter for the next couple of weeks. You may find that additional review is delayed until 2024. Thanks again for submitting this PR!

dev-rahulbhadoriya commented:
@matthew-white , thanks for the suggestions and for reviewing my PR. I have corrected the code and added a test case for it. Kindly check and review it

matthew-white commented on test/integration/api/users.js:
```diff
@@ -402,6 +402,12 @@ describe('api: /users', () => {
               .send({ new: 'coolpassword' })
               .expect(403))));
+
+        it('should fail the request if email field is sent blank in request body', testService((service) =>
+          service.login('alice', (asAlice) =>
+            asAlice.post('/v1/users/reset/initiate')
+              .send({ email: '' })
+              .expect(400))));
```
Would you mind also adding assertions about the response body? `code` should equal 400.2, and `details` should equal `{ field: 'email' }`.

dev-rahulbhadoriya replied:
can do something like this ?
`it('should fail the request if email field is sent blank in request body', testService((service) =>
          service.login('alice', (asAlice) =>
            asAlice.post('/v1/users/reset/initiate')
              .send({ email: '' })
              .expect(400)
              .then(({ body: { code, details } }) => {
                details.should.eql({ field: 'email' });
                code.should.eql(400.2);
              }))));`

matthew-white reviewed (approved):
Sorry for the long delay on review! This is looking great. I'll go ahead and merge now. Thank you for the PR, @dev-rahulbhadoriya!
